File: python/path-sum-iii.py
import collections

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, s: int) -> int:
        def dfs(node, prev_sum, dic):
            if not node: return 0
            
            curr_sum = prev_sum + node.val
            res = dic[curr_sum - s]
            dic[curr_sum] += 1
            res += dfs(node.left, curr_sum, dic)
            res += dfs(node.right, curr_sum, dic)
            
            dic[curr_sum] -= 1
            return res
        
        return dfs(root, 0, collections.defaultdict(int, {0: 1}))

File: go/path_sum_iii.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) int {
	dic := make(map[int]int)
	dic[0] = 1
	return dfs(root, 0, targetSum, dic)
}

func dfs(node *TreeNode, prev_sum int, t_sum int, dic map[int]int) int {
	if node == nil {
		return 0
	}

	curr_sum := prev_sum + node.Val
	res := dic[curr_sum-t_sum]
	dic[curr_sum] += 1

	res += dfs(node.Left, curr_sum, t_sum, dic)
	res += dfs(node.Right, curr_sum, t_sum, dic)

	dic[curr_sum] -= 1
	return res
}
